File: java/client/src/test/java/org/signal/libsignal/net/UnauthMessagesServiceTest.kt
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.net

import kotlinx.serialization.json.Json
import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Test
import org.signal.libsignal.internal.TokioAsyncContext
import org.signal.libsignal.protocol.ServiceId.Aci
import org.signal.libsignal.zkgroup.groupsend.GroupSendFullToken
import java.nio.charset.StandardCharsets
import java.util.UUID
import java.util.concurrent.Future
import kotlin.io.encoding.Base64
import kotlin.test.assertIs

class UnauthMessagesServiceTest {
  private val recipientUuid = UUID.fromString("4FCFE887-A600-40CD-9AB7-FD2A695E9981")

  // From `SERIALIZED_GROUP_SEND_TOKEN` in Rust.
  private val testGroupSendToken =
    GroupSendFullToken(
      Base64.decode("ABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABo5c+LAQAA"),
    )

  private fun sendTestMultiRecipientMessage(
    chat: UnauthenticatedChatConnection,
    fakeRemote: FakeChatRemote,
  ): Pair<Future<RequestResult<MultiRecipientMessageResponse, MultiRecipientSendFailure>>, Long> {
    val messagesService = UnauthMessagesService(chat)

    val testPayload = byteArrayOf(1, 2, 3, 4)
    val timestamp = 1700000000000L
    val responseFuture =
      messagesService.sendMultiRecipientMessage(
        testPayload,
        timestamp,
        MultiRecipientSendAuthorization.Story,
        onlineOnly = false,
        urgent = true,
      )

    // Get the incoming request from the fake remote
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get()

    assertEquals("PUT", request.method)
    val expectedPath = "/v1/messages/multi_recipient?ts=1700000000000&online=false&urgent=true&story=true"
    assertEquals(expectedPath, request.pathAndQuery)
    assertEquals(mapOf("content-type" to "application/vnd.signal-messenger.mrm"), request.headers)
    assertArrayEquals(testPayload, request.body)

    return Pair(responseFuture, requestId)
  }

  @Test
  fun testSendMultiRecipientMessageSuccess() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val (responseFuture, requestId) = sendTestMultiRecipientMessage(chat, fakeRemote)

    // Send successful response with one unregistered UUID
    val jsonResponse =
      """
      {
        "uuids404": ["$recipientUuid"]
      }
      """.trimIndent()

    fakeRemote.sendResponse(
      requestId,
      200,
      "OK",
      arrayOf("content-type: application/json"),
      jsonResponse.toByteArray(StandardCharsets.UTF_8),
    )

    // Verify the result
    val result = responseFuture.get()
    val successResult = assertIs<RequestResult.Success<MultiRecipientMessageResponse>>(result)
    assertEquals(listOf(Aci(recipientUuid)), successResult.result.unregisteredIds)
  }

  @Test
  fun testSendMultiRecipientMessageUnauthorized() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val (responseFuture, requestId) = sendTestMultiRecipientMessage(chat, fakeRemote)
    fakeRemote.sendResponse(
      requestId,
      401,
      "Unauthorized",
      arrayOf(),
      byteArrayOf(),
    )

    // Verify the result
    val result = responseFuture.get()
    val nonSuccessResult = assertIs<RequestResult.NonSuccess<MultiRecipientSendFailure>>(result)
    assertIs<RequestUnauthorizedException>(nonSuccessResult.error)
  }

  @Test
  fun testSendMultiRecipientMessageMismatchedDevices() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val (responseFuture, requestId) = sendTestMultiRecipientMessage(chat, fakeRemote)

    val jsonResponse =
      """
      [
        {
          "uuid": "$recipientUuid",
          "devices": {
            "missingDevices": [4, 5],
            "extraDevices": [40, 50]
          }
        }
      ]
      """.trimIndent()

    fakeRemote.sendResponse(
      requestId,
      409,
      "Conflict",
      arrayOf("content-type: application/json"),
      jsonResponse.toByteArray(StandardCharsets.UTF_8),
    )

    // Verify the result
    val result = responseFuture.get()
    val nonSuccessResult = assertIs<RequestResult.NonSuccess<MultiRecipientSendFailure>>(result)
    val mismatchedDevices = assertIs<MismatchedDeviceException>(nonSuccessResult.error)
    assertArrayEquals(
      arrayOf(
        MismatchedDeviceException.Entry(
          Aci(recipientUuid),
          missingDevices = intArrayOf(4, 5),
          extraDevices = intArrayOf(40, 50),
        ),
      ),
      mismatchedDevices.entries,
    )
  }

  @Test
  fun testSendMultiRecipientMessageStaleDevices() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val (responseFuture, requestId) = sendTestMultiRecipientMessage(chat, fakeRemote)

    val jsonResponse =
      """
      [
        {
          "uuid": "$recipientUuid",
          "devices": {
            "staleDevices": [4, 5]
          }
        }
      ]
      """.trimIndent()

    fakeRemote.sendResponse(
      requestId,
      410,
      "Gone",
      arrayOf("content-type: application/json"),
      jsonResponse.toByteArray(StandardCharsets.UTF_8),
    )

    // Verify the result
    val result = responseFuture.get()
    val nonSuccessResult = assertIs<RequestResult.NonSuccess<MultiRecipientSendFailure>>(result)
    val mismatchedDevices = assertIs<MismatchedDeviceException>(nonSuccessResult.error)
    assertArrayEquals(
      arrayOf(
        MismatchedDeviceException.Entry(
          Aci(recipientUuid),
          staleDevices = intArrayOf(4, 5),
        ),
      ),
      mismatchedDevices.entries,
    )
  }

  @Test
  fun testSendMultiRecipientMessageServerSideError() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val (responseFuture, requestId) = sendTestMultiRecipientMessage(chat, fakeRemote)
    fakeRemote.sendResponse(
      requestId,
      500,
      "Internal Server Error",
      arrayOf(),
      byteArrayOf(),
    )

    // Verify the result
    val result = responseFuture.get()
    val retryableNetworkError = assertIs<RequestResult.RetryableNetworkError>(result)
    assertIs<ServerSideErrorException>(retryableNetworkError.networkError)
  }

  private fun sendTestSealedMessage(
    chat: UnauthenticatedChatConnection,
    auth: UserBasedSendAuthorization,
    expectedAuthHeader: Pair<String, String>?,
    fakeRemote: FakeChatRemote,
  ): Pair<Future<RequestResult<Unit, SealedSendFailure>>, Long> {
    val messagesService = UnauthMessagesService(chat)

    val timestamp = 1700000000000L
    val expectedBody =
      Json.parseToJsonElement(
        """
        {
          "messages": [
            {
              "type": 6,
              "destinationDeviceId": 1,
              "destinationRegistrationId": 11,
              "content": "AQID"
            },
            {
              "type": 6,
              "destinationDeviceId": 2,
              "destinationRegistrationId": 22,
              "content": "BAUG"
            }
          ],
          "online": false,
          "urgent": true,
          "timestamp": 1700000000000
        }
        """,
      )

    val responseFuture =
      messagesService.sendMessage(
        Aci(recipientUuid),
        timestamp,
        listOf(
          SingleOutboundSealedSenderMessage(1, 11, byteArrayOf(1, 2, 3)),
          SingleOutboundSealedSenderMessage(2, 22, byteArrayOf(4, 5, 6)),
        ),
        auth,
        onlineOnly = false,
        urgent = true,
      )

    // Get the incoming request from the fake remote
    val (request, requestId) = fakeRemote.getNextIncomingRequest().get()

    assertEquals("PUT", request.method)
    val expectedPath = "/v1/messages/$recipientUuid" + if (expectedAuthHeader == null) "?story=true" else ""
    assertEquals(expectedPath, request.pathAndQuery)
    assertEquals(listOfNotNull("content-type" to "application/json", expectedAuthHeader).toMap(), request.headers)
    assertEquals(expectedBody, Json.parseToJsonElement(request.body.toString(Charsets.UTF_8)))

    return Pair(responseFuture, requestId)
  }

  @Test
  fun testSendMessageSuccess() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    listOf(
      UserBasedSendAuthorization.Story to null,
      UserBasedAuthorization.AccessKey(ByteArray(16, { 0x0A })) to
        ("unidentified-access-key" to "CgoKCgoKCgoKCgoKCgoKCg=="),
      UserBasedAuthorization.GroupSend(testGroupSendToken) to
        ("group-send-token" to Base64.encode(testGroupSendToken.serialize())),
      UserBasedAuthorization.UnrestrictedUnauthenticatedAccess to
        ("unidentified-access-key" to "AAAAAAAAAAAAAAAAAAAAAA=="),
    ).forEach { (auth, expectedAuthHeader) ->
      val (responseFuture, requestId) = sendTestSealedMessage(chat, auth, expectedAuthHeader, fakeRemote)

      fakeRemote.sendResponse(
        requestId,
        200,
        "OK",
        arrayOf("content-type: application/json"),
        "{}".toByteArray(),
      )

      // Verify the result
      val result = responseFuture.get()
      assertIs<RequestResult.Success<Unit>>(result)
    }
  }

  @Test
  fun testSendMessageNotFound() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val (responseFuture, requestId) = sendTestSealedMessage(chat, UserBasedSendAuthorization.Story, null, fakeRemote)
    fakeRemote.sendResponse(
      requestId,
      404,
      "Not Found",
      arrayOf(),
      byteArrayOf(),
    )

    // Verify the result
    val result = responseFuture.get()
    val nonSuccessResult = assertIs<RequestResult.NonSuccess<SealedSendFailure>>(result)
    assertIs<ServiceIdNotFoundException>(nonSuccessResult.error)
  }

  @Test
  fun testSendMessageUnauthorized() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val (responseFuture, requestId) = sendTestSealedMessage(chat, UserBasedSendAuthorization.Story, null, fakeRemote)
    fakeRemote.sendResponse(
      requestId,
      401,
      "Unauthorized",
      arrayOf(),
      byteArrayOf(),
    )

    // Verify the result
    val result = responseFuture.get()
    val nonSuccessResult = assertIs<RequestResult.NonSuccess<SealedSendFailure>>(result)
    assertIs<RequestUnauthorizedException>(nonSuccessResult.error)
  }

  @Test
  fun testSendMessageMismatchedDevices() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val (responseFuture, requestId) = sendTestSealedMessage(chat, UserBasedSendAuthorization.Story, null, fakeRemote)

    val jsonResponse =
      """
      {
        "missingDevices": [4, 5],
        "extraDevices": [40, 50]
      }
      """.trimIndent()

    fakeRemote.sendResponse(
      requestId,
      409,
      "Conflict",
      arrayOf("content-type: application/json"),
      jsonResponse.toByteArray(StandardCharsets.UTF_8),
    )

    // Verify the result
    val result = responseFuture.get()
    val nonSuccessResult = assertIs<RequestResult.NonSuccess<SealedSendFailure>>(result)
    val mismatchedDevices = assertIs<MismatchedDeviceException>(nonSuccessResult.error)
    assertArrayEquals(
      arrayOf(
        MismatchedDeviceException.Entry(
          Aci(recipientUuid),
          missingDevices = intArrayOf(4, 5),
          extraDevices = intArrayOf(40, 50),
        ),
      ),
      mismatchedDevices.entries,
    )
  }

  @Test
  fun testSendMessageStaleDevices() {
    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val (responseFuture, requestId) = sendTestSealedMessage(chat, UserBasedSendAuthorization.Story, null, fakeRemote)

    val jsonResponse =
      """
      {
        "staleDevices": [4, 5]
      }
      """.trimIndent()

    fakeRemote.sendResponse(
      requestId,
      410,
      "Gone",
      arrayOf("content-type: application/json"),
      jsonResponse.toByteArray(StandardCharsets.UTF_8),
    )

    // Verify the result
    val result = responseFuture.get()
    val nonSuccessResult = assertIs<RequestResult.NonSuccess<SealedSendFailure>>(result)
    val mismatchedDevices = assertIs<MismatchedDeviceException>(nonSuccessResult.error)
    assertArrayEquals(
      arrayOf(
        MismatchedDeviceException.Entry(
          Aci(recipientUuid),
          staleDevices = intArrayOf(4, 5),
        ),
      ),
      mismatchedDevices.entries,
    )
  }
}

File: node/ts/test/chat/UnauthMessagesServiceTest.ts
//
// Copyright 2025 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { assert, config, expect, use } from 'chai';
import chaiAsPromised from 'chai-as-promised';
import { Buffer } from 'node:buffer';

import * as Native from '../../Native.js';
import * as util from '../util.js';
import {
  MultiRecipientMessageResponse,
  TokioAsyncContext,
  UnauthMessagesService,
} from '../../net.js';
import { connectUnauth } from './ServiceTestUtils.js';
import { ErrorCode, LibSignalErrorBase } from '../../Errors.js';
import { Aci } from '../../Address.js';
import { FakeChatRemote, InternalRequest } from '../../net/FakeChat.js';
import { GroupSendFullToken } from '../../zkgroup/index.js';

use(chaiAsPromised);

util.initLogger();
config.truncateThreshold = 0;

const recipientUuid = '4fcfe887-a600-40cd-9ab7-fd2a695e9981';

// From `SERIALIZED_GROUP_SEND_TOKEN` in Rust.
const testGroupSendToken = new GroupSendFullToken(
  Buffer.from('ABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABo5c+LAQAA', 'base64')
);

describe('UnauthMessagesService', () => {
  describe('multi-recipient messages', () => {
    async function sendTestMultiRecipientMessage(
      chat: UnauthMessagesService,
      fakeRemote: FakeChatRemote
    ): Promise<[Promise<MultiRecipientMessageResponse>, InternalRequest]> {
      const payload = Uint8Array.of(1, 2, 3, 4);
      const timestamp = 1700000000000;
      const responseFuture = chat.sendMultiRecipientMessage({
        payload,
        timestamp,
        auth: 'story',
        onlineOnly: false,
        urgent: true,
      });

      // Get the incoming request from the fake remote
      const request = await fakeRemote.assertReceiveIncomingRequest();

      expect(request.verb).to.eq('PUT');
      expect(request.path).to.eq(
        '/v1/messages/multi_recipient?ts=1700000000000&online=false&urgent=true&story=true'
      );
      expect(request.headers).to.deep.eq(
        new Map([['content-type', 'application/vnd.signal-messenger.mrm']])
      );
      expect(request.body).to.deep.eq(payload);

      return [responseFuture, request];
    }

    it('can send', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthMessagesService>(tokio);

      const [responseFuture, request] = await sendTestMultiRecipientMessage(
        chat,
        fakeRemote
      );

      fakeRemote.sendReplyTo(request, {
        status: 200,
        message: 'OK',
        headers: ['content-type: application/json'],
        body: Buffer.from(
          JSON.stringify({
            uuids404: [recipientUuid],
          })
        ),
      });

      const responseFromServer = await responseFuture;
      assert(responseFromServer !== null);
      expect(responseFromServer.unregisteredIds).to.deep.equal([
        Aci.fromUuid(recipientUuid),
      ]);
    });

    it('can handle RequestUnauthorized', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthMessagesService>(tokio);

      const [responseFuture, request] = await sendTestMultiRecipientMessage(
        chat,
        fakeRemote
      );

      fakeRemote.sendReplyTo(request, {
        status: 401,
        message: 'Unauthorized',
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.RequestUnauthorized,
        });
    });

    it('can handle a mismatched device error', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthMessagesService>(tokio);

      const [responseFuture, request] = await sendTestMultiRecipientMessage(
        chat,
        fakeRemote
      );

      fakeRemote.sendReplyTo(request, {
        status: 409,
        message: 'Conflict',
        headers: ['content-type: application/json'],
        body: Buffer.from(
          JSON.stringify([
            {
              uuid: recipientUuid,
              devices: {
                missingDevices: [4, 5],
                extraDevices: [40, 50],
              },
            },
          ])
        ),
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.MismatchedDevices,
          entries: [
            {
              account: Aci.fromUuid(recipientUuid),
              missingDevices: [4, 5],
              extraDevices: [40, 50],
              staleDevices: [],
            },
          ],
        });
    });

    it('can handle a stale device error', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthMessagesService>(tokio);

      const [responseFuture, request] = await sendTestMultiRecipientMessage(
        chat,
        fakeRemote
      );

      fakeRemote.sendReplyTo(request, {
        status: 410,
        message: 'Gone',
        headers: ['content-type: application/json'],
        body: Buffer.from(
          JSON.stringify([
            {
              uuid: recipientUuid,
              devices: {
                staleDevices: [4, 5],
              },
            },
          ])
        ),
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.MismatchedDevices,
          entries: [
            {
              account: Aci.fromUuid(recipientUuid),
              missingDevices: [],
              extraDevices: [],
              staleDevices: [4, 5],
            },
          ],
        });
    });

    it('can handle server-side errors', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthMessagesService>(tokio);

      const [responseFuture, request] = await sendTestMultiRecipientMessage(
        chat,
        fakeRemote
      );

      fakeRemote.sendReplyTo(request, {
        status: 500,
        message: 'Internal Server Error',
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.IoError,
        });
    });
  });

  describe('1:1 messages', () => {
    async function sendTestSealedMessage(
      chat: UnauthMessagesService,
      auth:
        | 'story'
        | { accessKey: Uint8Array<ArrayBuffer> }
        | GroupSendFullToken
        | 'unrestricted',
      expectedAuthHeader: Readonly<[string, string]> | null,
      fakeRemote: FakeChatRemote
    ): Promise<[Promise<void>, InternalRequest]> {
      const timestamp = 1700000000000;
      const responseFuture = chat.sendMessage({
        destination: Aci.fromUuid(recipientUuid),
        timestamp,
        auth,
        contents: [
          { deviceId: 1, registrationId: 11, contents: Uint8Array.of(1, 2, 3) },
          { deviceId: 2, registrationId: 22, contents: Uint8Array.of(4, 5, 6) },
        ],
        onlineOnly: false,
        urgent: true,
      });

      // Get the incoming request from the fake remote
      const request = await fakeRemote.assertReceiveIncomingRequest();

      expect(request.verb).to.eq('PUT');
      expect(request.path).to.eq(
        `/v1/messages/${recipientUuid}${auth === 'story' ? '?story=true' : ''}`
      );
      expect(request.headers).to.deep.eq(
        new Map([
          ['content-type', 'application/json'],
          ...(expectedAuthHeader ? [expectedAuthHeader] : []),
        ])
      );
      expect(JSON.parse(new TextDecoder().decode(request.body))).to.deep.eq({
        messages: [
          {
            type: 6,
            destinationDeviceId: 1,
            destinationRegistrationId: 11,
            content: 'AQID',
          },
          {
            type: 6,
            destinationDeviceId: 2,
            destinationRegistrationId: 22,
            content: 'BAUG',
          },
        ],
        online: false,
        urgent: true,
        timestamp: 1700000000000,
      });

      return [responseFuture, request];
    }

    it('can send', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthMessagesService>(tokio);

      for (const [auth, expectedAuthHeader] of [
        ['story', null] as const,
        [
          { accessKey: util.repeatedBytes(0x0a, 16) },
          ['unidentified-access-key', 'CgoKCgoKCgoKCgoKCgoKCg=='],
        ] as const,
        [
          testGroupSendToken,
          [
            'group-send-token',
            Buffer.from(testGroupSendToken.serialize()).toString('base64'),
          ],
        ] as const,
        [
          'unrestricted',
          ['unidentified-access-key', 'AAAAAAAAAAAAAAAAAAAAAA=='],
        ] as const,
      ]) {
        const [responseFuture, request] = await sendTestSealedMessage(
          chat,
          auth,
          expectedAuthHeader,
          fakeRemote
        );

        fakeRemote.sendReplyTo(request, {
          status: 200,
          message: 'OK',
          headers: ['content-type: application/json'],
          body: Buffer.from('{}'),
        });

        await responseFuture;
      }
    });

    it('can handle RequestUnauthorized', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthMessagesService>(tokio);

      const [responseFuture, request] = await sendTestSealedMessage(
        chat,
        'story',
        null,
        fakeRemote
      );

      fakeRemote.sendReplyTo(request, {
        status: 401,
        message: 'Unauthorized',
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.RequestUnauthorized,
        });
    });

    it('can handle a mismatched device error', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthMessagesService>(tokio);

      const [responseFuture, request] = await sendTestSealedMessage(
        chat,
        'story',
        null,
        fakeRemote
      );

      fakeRemote.sendReplyTo(request, {
        status: 409,
        message: 'Conflict',
        headers: ['content-type: application/json'],
        body: Buffer.from(
          JSON.stringify({
            missingDevices: [4, 5],
            extraDevices: [40, 50],
          })
        ),
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.MismatchedDevices,
          entries: [
            {
              account: Aci.fromUuid(recipientUuid),
              missingDevices: [4, 5],
              extraDevices: [40, 50],
              staleDevices: [],
            },
          ],
        });
    });

    it('can handle a stale device error', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthMessagesService>(tokio);

      const [responseFuture, request] = await sendTestSealedMessage(
        chat,
        'story',
        null,
        fakeRemote
      );

      fakeRemote.sendReplyTo(request, {
        status: 410,
        message: 'Gone',
        headers: ['content-type: application/json'],
        body: Buffer.from(
          JSON.stringify({
            staleDevices: [4, 5],
          })
        ),
      });

      await expect(responseFuture)
        .to.eventually.be.rejectedWith(LibSignalErrorBase)
        .and.deep.include({
          code: ErrorCode.MismatchedDevices,
          entries: [
            {
              account: Aci.fromUuid(recipientUuid),
              missingDevices: [],
              extraDevices: [],
              staleDevices: [4, 5],
            },
          ],
        });
    });
  });
});
